lib/spack/spack/solver/asp.py

```diff
@@ -2240,10 +2238,7 @@ def _reusable_specs(self):
             # Specs from buildcaches
             try:
                 index = spack.binary_distribution.update_cache_and_get_specs()
-                reusable_specs.extend([
-                    s for s in index if not s.satisfies('dev_path=*')
```

Comment:
@becker33 @tgamblin Why did we have this check? Can it ever happen to have a dev spec in a binary index?

Reply:
Anyhow, with the latest modifications at 91890dc:
```console
$ time spack solve zlib
==> Best of 3 considered solutions.
==> Optimization Criteria:
  Priority  Criterion                                            Installed  ToBuild
  1         number of input specs not concretized                        -        0
  2         number of packages to build (vs. reuse)                      -        0
  3         deprecated versions used                                     1        0
  4         version weight                                               1        0
  5         number of non-default variants (roots)                       0        0
  6         preferred providers for roots                                0        0
  7         default values of variants not being used (roots)            0        0
  8         number of non-default variants (non-roots)                   0        0
  9         preferred providers (non-roots)                              0        0
  10        compiler mismatches                                          0        0
  11        OS mismatches                                                0        0
  12        non-preferred OS's                                           0        0
  13        version badness                                              1        0
  14        default values of variants not being used (non-roots)        0        0
  15        non-preferred compilers                                    100        0
  16        target mismatches                                            0        0
  17        non-preferred targets                                       16        0

zlib@1.2.11%gcc@9.3.0+optimize+pic+shared arch=linux-ubuntu20.04-x86_64


real	0m20,380s
user	0m18,827s
sys	0m0,408s
```

Reply:
Discussed offline with @becker33 It seems it's fine to expect that dev specs won't be in build caches.